rig/place_and_route/utils.py

```diff
@@ -102,8 +106,29 @@ def build_routing_tables(routes, net_keys, omit_default_routes=True):
                     out_directions.add(child)
 
             # Add a routing entry when the direction changes
-            if not omit_default_routes or set([direction]) != out_directions:
-                routing_tables[(x, y)].append(
-                    RoutingTableEntry(out_directions, key, mask))
+            if (key, mask) in route_sets[(x, y)]:
+                # Update the existing route set if possible
+                route_sets[(x, y)][(key, mask)].ins.add(direction)
+                route_sets[(x, y)][(key, mask)].outs.update(out_directions)
+            else:
+                # Otherwise create a new route set
+                route_sets[(x, y)][(key, mask)] = _InOutPair(
+                    set([direction]), set(out_directions)
+                )
+
+    # Construct the routing tables from the route sets
+    routing_tables = defaultdict(list)
+    for (x, y), routes in iteritems(route_sets):
+        for (key, mask), route in iteritems(routes):
+            # Remove default routes where possible
+            if omit_default_routes and (len(route.ins) == 1 and
+                                        route.outs == route.ins):
```

Comment:
`ins` comes from `direction` (line 91 in this commit) which appears to be the direction out of the previous node.  (I copied you here).